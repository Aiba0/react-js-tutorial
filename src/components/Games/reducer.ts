import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { lensPath, set } from "ramda";

import { Coordinates } from "@/rdx/actions";

import { createEmptyGameField } from "./fieldManager";

export enum GameStatus {
  NewGame = "New game",
  Play = "Play",
  GameOver = "Game over",
}

export type FieldSizeType = [number, number];
export type PlayerMarksType = [string, string];
export type GameFieldType = string[][];

export const PlayerMarks: PlayerMarksType = ["x", "o"];
export const [firstPlayerMark, secondPlayerMark] = PlayerMarks;

export const defaultFieldSize: FieldSizeType = [3, 3];
const [defaultSizeX, defaultSizeY] = defaultFieldSize;

export type RebuildActionType = PayloadAction<{
  fieldSize?: FieldSizeType;
  playerMarks?: PlayerMarksType;
  nextTurn?: string;
}>;

export type ClickActionType = PayloadAction<Coordinates>;
export type ChangeStatusActionType = PayloadAction<GameStatus>;
export type SetWinnerActionType = PayloadAction<string>;

export const initialState: {
  fieldSize: FieldSizeType;
  playerMarks: PlayerMarksType;
  gameField: GameFieldType;
  gameStatus: GameStatus;
  nextTurn: string;
  moves: number;
  winner?: string;
} = {
  fieldSize: defaultFieldSize,
  playerMarks: PlayerMarks,
  gameField: createEmptyGameField(defaultSizeX, defaultSizeY),
  gameStatus: GameStatus.NewGame,
  nextTurn: firstPlayerMark,
  moves: 0,
  winner: undefined,
};

export const gameSlice = createSlice({
  name: "game",
  initialState,
  reducers: {
    rebuild: (state) => ({
      ...state,
      ...initialState,
    }),
    click: (state, { payload }: ClickActionType) => {
      const { x, y } = payload;
      const { nextTurn, gameField, moves } = state;
      return {
        ...state,
        gameField: set(lensPath([y, x]), nextTurn, gameField),
        nextTurn:
          nextTurn === firstPlayerMark ? secondPlayerMark : firstPlayerMark,
        moves: moves + 1,
      };
    },
    changeStatus: (state, { payload }: ChangeStatusActionType) => ({
      ...state,
      gameStatus: payload,
    }),
    setWinner: (state, { payload }: SetWinnerActionType) => ({
      ...state,
      winner: payload,
    }),
  },
});

export const { reducer, actions } = gameSlice;
