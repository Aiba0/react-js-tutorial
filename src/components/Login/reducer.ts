import { createSlice } from "@reduxjs/toolkit";

export const initialState = { username: "" };

export const loginSlice = createSlice({
  name: "user",
  initialState,
  reducers: {
    setUsername: (state, { payload }) => {
      if (payload.length > 3) {
        return { username: payload };
      }
      return state;
    },
  },
});
