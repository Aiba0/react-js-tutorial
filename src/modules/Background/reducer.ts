import { createSlice } from "@reduxjs/toolkit";

import { TicTacToeGameState } from "@/store";

export const BackgroundStatus = {
  init: "init",
  start: "start",
  cancel: "cancel",
};

export const selectors = {
  background: ({ background }: TicTacToeGameState) => background,
};

export const initialState = {
  status: BackgroundStatus.init,
};

export const backgroundSlice = createSlice({
  name: "background",
  initialState,
  reducers: {
    start: (state) => {
      state.status = BackgroundStatus.start;
      return state;
    },
    cancel: (state) => {
      state.status = BackgroundStatus.cancel;
      return state;
    },
  },
});

export const { reducer, actions } = backgroundSlice;
