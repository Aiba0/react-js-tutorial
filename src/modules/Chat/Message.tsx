import React, { FC } from "react";

interface Props {
  author: string;
  children: string;
}

export const MessageComponent: FC<Props> = ({ author, children }) => (
  <div>
    <b>{author}: </b>
    <span>{children}</span>
  </div>
);
