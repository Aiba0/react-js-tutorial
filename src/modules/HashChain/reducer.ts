import { createSlice } from "@reduxjs/toolkit";

export const initialState = {
  hash: "",
};

export const hashSlice = createSlice({
  name: "hash",
  initialState,
  reducers: {
    update: (state, { payload }) => ({ hash: payload }),
  },
});

export const { reducer, actions } = hashSlice;
