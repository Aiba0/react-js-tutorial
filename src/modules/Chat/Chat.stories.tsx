import React from "react";
import { withKnobs, text } from "@storybook/addon-knobs";
import { action } from "@storybook/addon-actions";

import { ChatComponent } from "./Chat";
import { actions } from "./reducer";

export default {
  title: "ChatComponent",
  decorators: [withKnobs],
};

export const ChatComponentExample = () => (
  <ChatComponent
    chat={[{ author: "N", message: "Test" }]}
    username={text("A", "N")}
    send={action("SendAction") as typeof actions.send}
  />
);
