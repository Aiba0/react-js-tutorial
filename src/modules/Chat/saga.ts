import { take, takeEvery, put, call } from "redux-saga/effects";
import { eventChannel } from "redux-saga";
import SocketIO from "socket.io-client";

import { actions } from "./reducer";

export const createWebSocketConnection = () =>
  SocketIO("http://localhost:3000");

export function createSocketChannel(socket: SocketIOClient.Socket) {
  return eventChannel((emit) => {
    const chatMessageHandler = ({
      payload,
    }: ReturnType<typeof actions.send>) => {
      emit(payload);
    };

    socket.on("chatMessage", chatMessageHandler);

    const unsubscribe = () => {
      socket.off("chatMessage", chatMessageHandler);
    };

    return unsubscribe;
  });
}

export function* sendMessage(
  socket: SocketIOClient.Socket,
  payload: ReturnType<typeof actions.send>
) {
  socket.emit("chatMessage", payload);
}

export function* chatSaga() {
  const socket = yield call(createWebSocketConnection);
  const socketChannel = yield call(createSocketChannel, socket);

  yield takeEvery(actions.send.type, sendMessage, socket);

  while (true) {
    try {
      const payload = yield take(socketChannel);
      yield put(actions.update([payload]));
    } catch (err) {
      console.error("socket error:", err);
      socketChannel.close();
    }
  }
}
