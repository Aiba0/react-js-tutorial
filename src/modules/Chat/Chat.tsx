import React, { useCallback, useState } from "react";
import { connect } from "react-redux";
import { isEmpty } from "ramda";

import { TicTacToeGameState } from "@/store";

import { MessageComponent } from "./Message";
import { actions } from "./reducer";

const mapStateToProps = ({ chat, login }: TicTacToeGameState) => ({
  chat,
  ...login,
});

const mapDispatchToProps = {
  send: actions.send,
};

export type Props = ReturnType<typeof mapStateToProps> &
  typeof mapDispatchToProps;

export const ChatComponent: React.FC<Props> = ({ chat, username, send }) => {
  const [message, setMsg] = useState("");
  const onSubmit = useCallback(
    async (ev) => {
      ev.preventDefault();
      if (!isEmpty(message)) {
        send({ message, author: username });
      }
    },
    [message, username, send]
  );
  return !isEmpty(username) ? (
    <>
      {chat.map(({ author, message }, idx) => (
        <MessageComponent key={`${author}_${message}_${idx}`} author={author}>
          {message}
        </MessageComponent>
      ))}
      <form onSubmit={onSubmit}>
        <label>
          Message:
          <input
            placeholder="Enter your msg"
            value={message}
            onChange={(ev) => setMsg((ev.target as HTMLInputElement).value)}
            required
            minLength={4}
            maxLength={10}
          />
        </label>
        <button>Send</button>
      </form>
    </>
  ) : null;
};

export const Chat = connect(mapStateToProps, mapDispatchToProps)(ChatComponent);
