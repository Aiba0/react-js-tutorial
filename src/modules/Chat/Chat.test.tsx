import React from "react";
import { shallow } from "enzyme";
import faker from "faker";

import { actions } from "./reducer";
import { ChatComponent } from "./Chat";

const props = {
  chat: [],
  username: faker.name.firstName(),
  send: actions.send,
};

describe("ChatComponent test", () => {
  it("Render chat when username is not empty", () => {
    const component = <ChatComponent {...props} />;

    expect(shallow(component).html()).toMatchSnapshot();
  });
  it("Render null when username is empty", () => {
    const component = <ChatComponent {...{ ...props, username: "" }} />;

    expect(shallow(component).html()).toMatchSnapshot();
  });
});
