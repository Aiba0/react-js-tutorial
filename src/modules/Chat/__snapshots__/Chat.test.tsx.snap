// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatComponent test Render chat when username is not empty 1`] = `"<form><label>Message:<input placeholder=\\"Enter your msg\\" value=\\"\\" required=\\"\\" minLength=\\"4\\" maxLength=\\"10\\"/></label><button>Send</button></form>"`;

exports[`ChatComponent test Render null when username is empty 1`] = `null`;
